### DISCUSSION TITLE
how to keep larger size of overlayed (composited) images

### DISCUSSION THREAD

**Discussion by grenzionky** [ImageMagick/ImageMagick#2147] [Help] [answered]
Im trying to get ImageMagick to keep the size of the larger image when compositing.

So whether i do `convert -background none -pointsize 36 rose: label:A -gravity center -composite out.png` or `convert -background none -pointsize 36 label:A rose: -gravity center -composite out.png` the result should look the same.

This is what it does by default:
![out](https://user-images.githubusercontent.com/13357949/84682017-014fef80-af03-11ea-9218-83e1b3ac7261.png)
`convert -background none -pointsize 36 rose: label:A -gravity center -composite out.png`
![out](https://user-images.githubusercontent.com/13357949/84682041-0c0a8480-af03-11ea-8f21-af7180f4b684.png)
`convert -background none -pointsize 36 label:A rose: -gravity center -composite out.png`

I want both versions to look like the first image.
Is there a way to do this?

**Comment by snibgo** [comment]
At "-composite", the output size will come from the size of its _first_ input. You can control the direction of the compose, by "-compose SrcOver" (puts pixels of second over pixels of first) or "-compose DstOver" (puts first over second).

"-compose Over" is the same as "-compose SrcOver".

I can't see a simple way of telling IM to put the largest image first. It can be done in a messy way, based on an expression in "-delete %[fx:...], so you read or create two images then delete the smallest.

**Comment by fmw42** [comment, marked as answer]
How about just using +swap

`convert -background none -pointsize 36 label:A rose: +swap -gravity center -composite out.png`

![out](https://user-images.githubusercontent.com/13458218/84688242-ddd17880-aef3-11ea-8aae-b9f2f0175f4c.png)

**Reply by grenzionky** [reply]
how do i use `+swap` only when the first image is the smaller one?

**Reply by fmw42** [reply]
You would have to test the size of the two images and use a variable that would be +swap if one way and and empty variable if the other way.

How do you decide which is bigger if one is larger in one dimension and smaller in the other dimension?

What is your IM version and platform?  If on unix-like, then you can do

```
test=$(convert -background none -pointsize 36 label:A rose: -format "%[fx:max(u.w,u.h)<max(v.w,v.h)?1:0]" info:)
if [ test -eq 1 ]; then
var="+swap"
else
var=""
fi
convert -background none -pointsize 36 label:A rose: $var -gravity center -composite out.png
```

**Reply by grenzionky** [reply]
the version you sent doesnt work for some reason, but it gave me a good headstart. thanks. heres my working version:

```
first=i
second=w
test=$(convert -background none -pointsize 36 label:"$first" label:"$second" -format "%[fx:u.w<v.w?1:0]" info:); if [ $test -eq 11 ]; then var="+swap"; else var=""; fi
convert -background none -pointsize 36 label:$first label:$second $var -gravity center -composite out.png
```

not sure why max() doesnt work,but its fine for my specific use case since i really only care about the width

**Reply by fmw42** [reply]
Sorry, here is corrected code that tests only on width:

```
test=$(convert -background none -pointsize 36 label:A rose: -format "%[fx:u.w<v.w?1:0]\n" info: | head -n 1)
if [ $test -eq 1 ]; then
var="+swap"
else
var=""
fi
convert -background none -pointsize 36 label:A rose: $var -gravity center -composite out.png

```

**Comment by snibgo** [comment]
Yes, that should work, but does almost everything twice.

It can be done in one v7 command, reading each image only once. Windows BAT syntax:
```
magick ^
  in0.png in1.png ( -clone 0 ) ^
  -delete "%%[fx:u.w>v.w?2:0]" ^
  -compose Over -composite ^
  out.png
```

**Reply by fmw42** [reply]
Nice concept, but only for IM 7

**Reply by grenzionky** [reply]
heres my working unix version:
```
magick -background none -pointsize 36 label:i label:w \( -clone 0 \) -delete "%[fx:u.w>v.w?2:0]" -gravity center -compose Over -composite out.png
```
i probably didnt do it correctly, but it works, so im happy 🤷

**Reply by fmw42** [reply]
Looks OK to me.

**Reply by grenzionky** [reply]
is there a way to make the background of the composited image white instead of transparent?

**Reply by fmw42** [reply]
add -background white -flatten after -composite and before saving the file
